[DSCP-431] use student API in faircv tool

pasqu4le opened the pull request:
### Description

This PR aims at using the student API in the tool that produces faircvs, to better test these interactions.

The WIP label has been removed because this PR also contains the fixes for DSCP-450, the issue tracking the problem that this PR itself revealed.

### YT issue

https://issues.serokell.io/issue/DSCP-431

### Checklist

Hint: a perfect PR has all the checkmarks set.

Possible related changes (conditional):
- [x] I added tests if required, in case they are:
  - Covering new introduced functionality (feature).
  - Regression tests preventing the bug from silently reappearing again (bugfix).
- [x] I have checked the [sample config](../tree/master/docs/config-full-sample.yaml) and [launch scripts](../tree/master/scripts/launch) and updated them if it was necessary (especially if executables or CLIs were changed).
- [x] I have checked READMEs and changed them accordingly if it's necessary (the main [README.md](../tree/master/README.md) as well as subproject READMEs for all related executables).
- [x] If a public API structure or/and data serialization was changed, I made sure that these changes are reflected in the:
  - [Swagger](../tree/master/specs/disciplina) API specs.
  - Any [related documentation](../tree/master/docs/api-types.md).
- [x] If any public documentation was written, I have spellchecked it.

Stylistic (obligatory):
- [x] My commit history is clean, descriptive and do not contain merge or revert commits or commits like "WIP".
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).
- [x] My code is documented (haddock) and these docs are decent (full enough and spellchecked).

pasqu4le committed:
[DSCP-431] use student API in faircv tool

Martoon-00 committed:
Fix transactions ordering in `getProofs`

Problem: `getProvenTransactions` function returned transactions with their indices within
block and without any
ordering, and in `getProofs` endpoint of Student API I just ignored indices and returned
transactions in 'some' order. But if we use those proofs to construct a fair CV then order
definitely matters. I suppose it matters in general as well, who would like to receive
transactions in arbitrary order at frontend :shrug:.

Solution: instead of returning transaction indices I sort them by database engine means.

pasqu4le committed:
[DSCP-431] improve proof waiting, add options

Martoon-00 commented on scripts/launch/node.sh:
```diff
@@ -139,25 +139,25 @@ fi
 elif [[ "$node" == "faucet" ]]; then
     if [ -z "$(which dscp-faucet)" ]; then
-        stack exec "dscp-faucet" -- $common_params $faucet_params
+        stack exec "dscp-faucet" --nix -- $common_params $faucet_params
```
Please do not do this. Not everyone uses nix.
I believe that this can be set somehow globally on your machine, maybe in `.stack/global-project/stack.yaml`.

pasqu4le replied:
sorry that was not intended, I changed it for a quick run and forgot about it, I'll remove it

Martoon-00 commented on tools/src/faircv/Main.hs:
```diff
@@ -76,15 +76,22 @@ main = do
+    -- Wait for the proofs to be available, then get them and make a faircv
     logInfo "Submissions sent, waiting for proofs"
-    threadDelay $ sec 20
+    waitForProofs subInfos
```
:cool:

Martoon-00 replied:
IMO it may worth waiting for a little bit more (fractions of seconds) after proofs are found, since your ask educator for proofs, but then verify FairCV in witness. It may take a little while for the educator to send its block to the witness.

Martoon-00 commented on tools/src/faircv/Options.hs:
```diff
@@ -0,0 +1,68 @@
+
+
+data FaircvOptions = FaircvOptions
+    { witnessUrl    :: String
```
What do you think about the following:
Make this parameter optional. In `node.sh` and `demo.sh` make `witness-listen` and `educator-listen` parameters have the same value - this will effectively make educator node serve both witness and educator APIs on the same port. And from this executable, you request `educatorUrl` when `witnessUrl` is unspecified.

pasqu4le replied:
I liked the idea, but it seems we cannot do that. I just tried and it seems that communications using the `StudentApiClient` result all in a 404.

Martoon-00 replied:
Weird, I have run the executable like `stack exec dscp-faircv localhost:8091 localhost:8090 ./run/tmp/educator.key` and that was working.
But ok, that's not significant.

pasqu4le replied:
well, but `8090` and `8091` are different, maybe I didn't understand your suggestion

Martoon-00 replied:
Generally, you need no cluster of witnesses, having a single educator (which is also a witness) is enough. So you can replace `8091` with `8090` [here](https://github.com/DisciplinaOU/disciplina/blob/8032e15454ab099030dfc9d83c6e8e2c1f6b1be1/scripts/launch/node.sh#L54), and launch

```
> ./scripts/launch/node.sh e b
> stack exec dscp-faircv -- -w localhost:8090 -e localhost:8090 -k ./run/tmp/educator.key
``` 
Now `-w` and `-e` accept the same value and it works

Although you may have reasons to launch separate witnesses especially since that discovered a bug.

pasqu4le replied:
Oh ok, I understood you then, but I have tried with a cluster (unsuccessfully).
I'll give it another try later.

Martoon-00 commented on tools/src/faircv/Main.hs:
```diff
@@ -1,26 +1,31 @@
+        return $ withIntSeed (fromIntegral sd) genSecretKey
+    , do
+        return $ withSeed input genSecretKey
+    ]
```
Would be very nice to move this (and the original function) somewhere to `Dscp.Crypto.?` with a note "this is how we generate secret keys from user-supplied seed".
I did this once, but then my change occurred to be redundant.

pasqu4le replied:
Done, I moved the function in `Dscp.Crypto.Util` which seemed a fair place.

Martoon-00 replied:
Thanks :+1:

Martoon-00 commented on tools/src/faircv/Options.hs:
```diff
@@ -0,0 +1,68 @@
+module Options
```
Try not to use such general name, `stack ghci` goes mad when it sees two modules with the same name (except for `Main`).

pasqu4le replied:
Done, I also renamed the Client module for the same reason.

pasqu4le committed:
[DSCP-431] rename Client and Options module to be more specific, abstract common secretFromSeed function, remove nix option from node script

pasqu4le committed:
[DSCP-431] remove faircv example

Martoon-00 committed:
[DSCP-431] Fix `getPrivateBlocksAfter`

Somehow I left a wrong comment in the function definition (with "including" instead of
"excluding"), and then rewrote everything on Beam implementing the function according
to its wrong definition.

A test on errorneous situation is added.

pasqu4le committed:
[DSCP-450] fix function call after rebase, add postgresql to shell-native.nix

Martoon-00 committed:
[DSCP-431] Use `relayTx` instead of `addTxToMempool` in publication

Our dataflow assumes `relayTx` usage when transaction should be submitted,
otherwise we do not submit transaction over network (until they appear in block).

Martoon-00 committed:
[DSCP-450] Fix in exceptions

pasqu4le committed:
[DSCP-450] deactivate 2 of educator's tests

pasqu4le committed:
[DSCP-431] add README documentation, add more options, add logging of informations

Martoon-00 commented on educator/tests/Test/Dscp/Educator/Publishing.hs:
```diff
@@ -15,8 +15,10 @@ import Test.Dscp.DB.SQL.Mode
-spec_Private_blocks_publishing :: Spec
-spec_Private_blocks_publishing = specWithTempPostgresServer $ do
+-- | This was renamed from "spec_Private_blocks_publishing" to be ignored.
+-- Both the tests it contains do not pass anymore and a solution is not known yet.
```
Please mention `DSCP-464` ticket here (I've just created it) so that we remember where this issue should be resolved.

Martoon-00 commented on tools/README.md:
````diff
@@ -127,3 +127,27 @@ echo 456 \
+```bash
+dscp-faircv -w 127.0.0.1:8013 -e 127.0.0.1:8090 -k run/tmp/educator/educator.key
+```
+
````
Cool :+1:

Martoon-00 commented on tools/src/faircv/Main.hs:
```diff
@@ -35,74 +31,70 @@ instance MonadLogging IO where
+    checkFairCV wClient fcv >>= \res ->
+        if fairCVFullyValid res
+        then logInfo "FairCV was successfully validated"
+        else logError "FairCV was not validated successfully"
```
Now it looks like a serious executable :+1:

Martoon-00 commented on educator/src/Dscp/Educator/Logic/Publishing.hs:
```diff
@@ -46,7 +46,7 @@ publishMissingBlocks =
 
         logInfo . fmt $ nameF "Adding publications to mempool" (listF $ toPtxId . ptwTx <$> pubTxs)
-        forM_ pubTxs $ addTxToMempool . GPublicationTxWitnessed
+        forM_ pubTxs $ relayTx . GPublicationTxWitnessed
```
Actually, I'm still wrong here. It should be 
```
forM_ pubTxs $ \pubTx -> relayTx (GPublicationTxWitnessed pubTx) >>= wait @"tx in mempool"
```

Otherwise, since `updateMempoolWithPublications` is invoked every second, on high load when transactions which are put to the relay does not appear in mempool immediately, we will effectively fill the relay with many duplicates of our publications.

@pasqu4le do you mind if I make this change and force-push?

pasqu4le replied:
@Martoon-00 no problem, please do.
Just please check that everything stay signed, because for some reason I seem to loose some signatures when rebasing or cherry-picking.
You could also add another commit on top of this, perhaps

flyingleafe replied:
@Martoon-00 why didn't you proceed with this change?

Martoon-00 commented on shell-native.nix:
```diff
@@ -7,6 +7,7 @@ stdenv.mkDerivation rec {
     binutils
     haskell.compiler.ghc822
     pkgconfig
+    postgresql
```
Sorry for this, I have seen this line in `default.nix` and was sure that was enough.

pasqu4le replied:
It's ok, also even with this `stack --nix` has some problems with tests (issue [DSCP-462](https://issues.serokell.io/issue/DSCP-462)).

P.s. I saw your comment on YouTrack, but haven't tried yet.

pasqu4le committed:
[DSCP-431] add mention to DSCP-464 in related comment

Martoon-00 commented on tools/src/faircv/FaircvOptions.hs:
```diff
@@ -0,0 +1,99 @@
+    help "File that will contain the resulting FairCV"
+
+contentSeedParser :: Parser Text
+contentSeedParser = option auto $
```
This is wrong btw, `auto` means that `read @Text` is used, and it expects strings like `"something"`.
I had to pass them like `-c "\"something\""` from shell.

I will fix this.

Martoon-00 committed:
[DSCP-431] Fix seed parsers in `faircv` executable

flyingleafe committed:
[DSCP-431] Fix faircv tool after Student API parameters update
